File: android/src/main/java/com/legichain/react/LegichainKycModule.kt
package com.legichain.react

import android.app.Activity
import android.content.Intent
import com.facebook.react.bridge.*
import com.facebook.react.ReactPackage
import com.facebook.react.uimanager.ViewManager
import com.legichain.kyc.KycOptions
import com.legichain.kyc.LegichainKyc
import org.json.JSONObject

class LegichainKycModule(private val context: ReactApplicationContext): ReactContextBaseJavaModule(context),ActivityEventListener {
    private var pending: Promise? = null
    init { context.addActivityEventListener(this) }
    override fun getName() = "LegichainKyc"
    @ReactMethod fun start(options: ReadableMap,promise: Promise) {
        val activity=currentActivity ?: run { promise.reject("NO_ACTIVITY","A foreground activity is required");return }
        if(pending!=null) { promise.reject("BUSY","A KYC flow is already running");return }
        try {
            val config=KycOptions(options.getString("apiToken").orEmpty(),options.getString("baseUrl") ?: "https://api.legichain.com",
                options.getString("language") ?: "tr",JSONObject(options.getMap("application")?.toHashMap() ?: emptyMap<String,Any>()))
            pending=promise
            activity.runOnUiThread {
                try { activity.startActivityForResult(LegichainKyc.intent(activity,config),REQUEST) }
                catch(e: Exception) { pending=null;promise.reject("START_FAILED","Could not open KYC") }
            }
        } catch(e: Exception) { pending=null;promise.reject("INVALID_OPTIONS","Invalid KYC options") }
    }
    override fun onActivityResult(activity: Activity,requestCode: Int,resultCode: Int,data: Intent?) {
        if(requestCode!=REQUEST) return
        val promise=pending ?: return;pending=null
        val value=JSONObject(LegichainKyc.result(data) ?: "{\"status\":\"cancelled\",\"application_id\":\"\"}")
        promise.resolve(Arguments.createMap().apply { putString("status",value.getString("status"));putString("application_id",value.optString("application_id")) })
    }
    override fun onNewIntent(intent: Intent) {}
    override fun invalidate() { pending?.reject("CANCELLED","React context closed");pending=null;context.removeActivityEventListener(this);super.invalidate() }
    companion object { const val REQUEST=46721 }
}

class LegichainKycPackage: ReactPackage {
    override fun createNativeModules(context: ReactApplicationContext): List<NativeModule> = listOf(LegichainKycModule(context))
    override fun createViewManagers(context: ReactApplicationContext): List<ViewManager<*,*>> = emptyList()
}

File: ios/LegichainKycModule.swift
import Foundation
import UIKit
import React

@objc(LegichainKyc)
final class LegichainKycModule: NSObject {
    private var active = false
    @objc static func requiresMainQueueSetup() -> Bool { true }
    @objc(start:resolver:rejecter:)
    func start(_ options: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            guard !self.active else { reject("BUSY","A KYC flow is already running",nil);return }
            guard let token=options["apiToken"] as? String,!token.isEmpty,
                  let url=URL(string:options["baseUrl"] as? String ?? "https://api.legichain.com"),url.scheme=="https",
                  let host=RCTPresentedViewController() else { reject("INVALID_OPTIONS","Invalid KYC configuration or presenter",nil);return }
            let config=KycOptions(apiToken:token,baseURL:url,language:options["language"] as? String ?? "tr",application:options["application"] as? [String:Any] ?? [:])
            let screen=KycViewController(options:config);self.active=true
            screen.onComplete={ result in self.active=false;resolve(result.dictionary) }
            host.present(screen,animated:true)
        }
    }
}
